### ISSUE TITLE
loss function

### ISSUE THREAD

**Issue by lihongxiacream** [Reason-Wang/ToolGen#6]
Thank you for open-sourcing such excellent work.I would like to ask about the loss function. When will the code for the loss functions used in the first and second rounds of training in your paper be open-sourced?Did you use Softmax to calculate the logits?

**Comment by lihongxiacream** [comment]
Additionally, I would like to ask how you handle queries without function calls, and how do you define this kind of output?Looking forward to your reply!!!!!!

**Comment by Reason-Wang** [comment]
Hi, thanks for your interest in our work. Our code is similar to Llama-Factory, which is just instruction finetuning. So the loss is just nex-token-prediction cross-entropy loss. You can simply take the first and second phase data as instruction data and train it using instruction tuning.

> When will the code be open-sourced?

Sorry for the delay. I am busy recently, and the code should be open-sourced within two weeks.

> Did you use Softmax to calculate the logits?

Yes. We simply use the transformers model to obtain token probabilities. And the probs are calculated from logits using Softmax.

> How to handle queries without function calls?

I am assuming you are asking queries that do not need function calls to solve. Our initial models are only trained with function call data, so I think it will always try to use function calls to solve problems. You can also try `reasonwang/ToolGen-WoSystem-Qwen2.5-3B-Instruct`, a model trained with mixture of instruction data and function call data. So it will decide whether to use function calls by itself.

**Comment by Reason-Wang** [comment]
Hi, I have just updated the training code in `training` folder.

**Comment by lihongxiacream** [comment]
I GOT IT . THANKS FOR YOUR REPLY!!
